conbench/entities/history.py

```diff
@@ -97,12 +87,73 @@ class HistorySerializer:
     many = _Serializer(many=True)
 
 
-def get_history(case_id: str, context_id: str, hardware_hash: str, repo: str) -> list:
-    """Given a case/context/hardware/repo, return all non-errored BenchmarkResults
-    (past, present, and future) on the default branch that match those criteria, along
-    with information about the stats of the distribution as of each BenchmarkResult.
-    Order is not guaranteed. Used to power the history API, which also powers the
-    timeseries plots.
+@dataclasses.dataclass
+class HistorySampleZscoreStats:
+    begins_distribution_change: bool
+    segment_id: str
+    rolling_mean_excluding_this_commit: float
+    rolling_mean: Optional[float]
+    residual: float
+    rolling_stddev: float
+
+
+@dataclasses.dataclass
+class HistorySample:
+    benchmark_result_id: str
+    case_id: str
+    context_id: str
+    # Is this `mean` really optional? When would this be None? this is
+    # populated from BenchmarkResult.mean which is nullable. It would be a
+    # major simplification if we knew that this is never None. Depends on the
+    # logic in BenchmarkResult.create() which is quite intransparent today,
+    # also there is a lack of spec
+    mean: Optional[float]
+    # math.nan is allowed for representing a failed iteration.
+    data: list[float]
+    times: list[float]
+    unit: str
+    hardware_hash: str
+    repository: str
+    commit_hash: str
+    commit_msg: str
+    # tz-naive timestamp of commit authoring time (to be interpreted in UTC
+    # timezone).
+    commit_timestamp: datetime.datetime
+    run_name: str
+    zscorestats: HistorySampleZscoreStats
+
+    def _dict_for_api_json(self) -> dict:
+        d = dataclasses.asdict(self)
+        # if performance is a concern then https://pypi.org/project/orjson/
+        # promises to be among the fastest for serializing python dataclass
+        # instances into JSON.
+        d["commit_timestamp"] = self.commit_timestamp.isoformat()
```

Comment:
timestamps/datetime objects can be represented in JSON in a variety of custom ways -- here we opt for a specific stringified notation. can review this in the future, was a quick decision.